### PR TITLE
[Workflow] Added Function (and Twig extension) to retrieve a specific transition

### PR DESCRIPTION
**Pull request by epitre**
You can now easily retrieve a specific transition. Useful when you want the metadata of a specific transition.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | hmmmm, should I create a ticket first ?
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13907

I needed to get the metadata of a transition, but in order to do that, you need the transition object. So here is a PR to easily get the transition object

### CODE REVIEW

**Comment by epitre** [comment]
Hi,
if someone can help me with the travis build failed.
I don't understand why it says that "Error: Call to undefined method Symfony\Component\Workflow\Workflow::getEnabledTransition()" as I did commit he method. You can see it here : https://github.com/symfony/symfony/pull/37428/files#diff-6d1e8dc0d2c0e7c4ab206d0efbb71516R241

and I don't understand all the other errors regarding src/Symfony/Component/Notifier/Bridge/FreeMobile, src/Symfony/Component/Notifier/Bridge/Slack, etc....
```PHP Fatal error:  Declaration of Symfony\Component\Notifier\Bridge\Slack\SlackTransport::doSend(Symfony\Component\Notifier\Message\MessageInterface $message): void must be compatible with Symfony\Component\Notifier\Transport\AbstractTransport::doSend(Symfony\Component\Notifier\Message\MessageInterface $message): Symfony\Component\Notifier\Message\SentMessage in /home/travis/build/symfony/symfony/src/Symfony/Component/Notifier/Bridge/Slack/SlackTransport.php on line 34```

I feel like it has nothing to do with this PR :/

**Comment by noniagriconomie** [comment]
For the notifier code, recently the return signature of the method changed

For the workflow part, i do not know :s

File: src/Symfony/Component/Workflow/Workflow.php [hunk 1]
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
...
 
+    /**
+     * {@inheritdoc}
+     */
+    public function getEnabledTransition(object $subject, string $name)
+    {
+        $marking = $this->getMarking($subject);
+
+        foreach ($this->definition->getTransitions() as $transition) {
+            $transitionBlockerList = $this->buildTransitionBlockerListForTransition($subject, $marking, $transition);
```
Comment (noniagriconomie) [inline src/Symfony/Component/Workflow/Workflow.php:246]: You can do the transition name equal comparison here to avoid useless computation to retreive the good one

File: src/Symfony/Component/Workflow/Workflow.php [hunk 2]
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
         return $enabledTransitions;
     }
 
+    /**
+     * {@inheritdoc}
```
Comment (noniagriconomie) [inline src/Symfony/Component/Workflow/Workflow.php:239]: It is not inside the interface thus no inherited doc here i fear

File: src/Symfony/Component/Workflow/Workflow.php [hunk 3]
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
...
+     */
+    public function getEnabledTransition(object $subject, string $name)
```
Comment (noniagriconomie) [inline src/Symfony/Component/Workflow/Workflow.php:241]: As a new method you can add return typehint i think here

File: src/Symfony/Component/Workflow/Workflow.php (hunk 1 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:246]: Done !
Thank you :)

File: src/Symfony/Component/Workflow/Workflow.php (hunk 2 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php]: you're right !
thank you :)

File: src/Symfony/Component/Workflow/Workflow.php (hunk 3 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php]: yes !
Thank you :)

File: src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php [hunk 4]
```diff
@@ -64,6 +65,16 @@ public function getEnabledTransitions(object $subject, string $name = null): arr
         return $this->workflowRegistry->get($subject, $name)->getEnabledTransitions($subject);
     }
 
+    /**
+     * Returns a transition or null.
+     *
+     * @return Transition|null Enabled transition asked for
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:71]: the docblock should be removed entirely

File: src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php [hunk 5]
```diff
@@ -64,6 +65,16 @@ public function getEnabledTransitions(object $subject, string $name = null): arr
...
+     */
+    public function getEnabledTransition(object $subject, string $transition, string $name = null): ?Transition
+    {
+        return $this->workflowRegistry->get($subject, $name)->getEnabledTransition($subject, $transition);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:75]: this means the bridge is now incompatible with workflow < 5.2, this should be reflected in the composer.json file, section "conflict" & "require-dev". That's the reason why deps=low tests fail.

Comment (epitre) [reply src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:70]: :pray:  Thank you.
I made the requested changes. Is it ok for you ?

I still have this error : ```1) Symfony\Component\Messenger\Bridge\Doctrine\Tests\Transport\DoctrineReceiverTest::testOccursRetryableExceptionFromConnection

Error: Class 'Doctrine\DBAL\Driver\PDOException' not found```

But I guess it has nothing to do with the Workflow component ^^

File: src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php (hunk 4 above)
Comment (epitre) [reply src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php]: Done !

File: src/Symfony/Component/Workflow/Tests/WorkflowTest.php [hunk 6]
```diff
@@ -592,6 +592,32 @@ public function testGetEnabledTransitions()
         $this->assertSame('t5', $transitions[0]->getName());
     }
 
+    public function testGetEnabledTransition()
+    {
+        $definition = $this->createComplexWorkflowDefinition();
+        $subject = new Subject();
+        $eventDispatcher = new EventDispatcher();
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Tests/WorkflowTest.php:599]: I don't think we need this complexity. Guard are already massively tested.
you can remove all things related to the Event Dispatcher + Guard?
It will make tests simpler to understand :) 
Thanks

File: src/Symfony/Component/Workflow/Workflow.php [hunk 7]
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
...
+    {
+        $marking = $this->getMarking($subject);
+
+        foreach ($this->definition->getTransitions() as $transition) {
+            if ($transition->getName() !== $name) {
+                continue;
+            }
+            $transitionBlockerList = $this->buildTransitionBlockerListForTransition($subject, $marking, $transition);
+            if ($transitionBlockerList->isEmpty()) {
+                return $transition;
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:248]: What if many transition match?
cf https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199

**Review by lyrixx — COMMENTED** [review]
Thanks for your contribution.

Instead of the current implementation, I would simplify a lot the code: 
You could leverage the current [`getEnabledTransitions()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199) method in the bridge, without implementing a new method.

Less code == less bug :)

**Comment by epitre** [comment]
> Thanks for your contribution.
> 
> Instead of the current implementation, I would simplify a lot the code:
> You could leverage the current [`getEnabledTransitions()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199) method in the bridge, without implementing a new method.
> 
> Less code == less bug :)

@lyrixx That's actually what I did first (and that's what I currently do in my project). And then I thought it would make more sense to have a dedicated function...
What I need in a twig template, someone might need it in a service (or somewhere else ^^)

File: src/Symfony/Component/Workflow/Workflow.php (hunk 7 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php:248]: :eyes: it's possible to have many transitions with the same name ? I did not realize...
In this case, I would have to add 2 optional arrays, one for the `froms` and one from the `tos`.
This way, it would be possible to identify specifically any transition.

I can try to do something, but I think it's overkill.

Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php]: So now, it is possible to retrieve any specific transition.
What do you think ?

File: src/Symfony/Component/Workflow/Tests/WorkflowTest.php (hunk 6 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Tests/WorkflowTest.php]: done !

**Comment by epitre** [comment]
Now, I have a problem with graphviz and PlantUML...

Maybe I should not have changed the complex workflow ^^

File: src/Symfony/Component/Workflow/Tests/WorkflowBuilderTrait.php [hunk 8]
```diff
@@ -19,6 +19,7 @@ private function createComplexWorkflowDefinition()
         $transitions[] = $transitionWithMetadataDumpStyle;
         $transitions[] = new Transition('t4', 'd', 'f');
         $transitions[] = new Transition('t5', 'e', 'g');
+        $transitions[] = new Transition('t3', 'f', 'e');
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Tests/WorkflowBuilderTrait.php:22]: As you state, IMHO, it's a bad idea to change theses fixtures :) You don't need it
Could you revert theses changes?

File: src/Symfony/Component/Workflow/Workflow.php [hunk 9]
```diff
@@ -235,6 +235,33 @@ public function getEnabledTransitions(object $subject)
         return $enabledTransitions;
     }
 
+    public function getEnabledTransition(object $subject, string $name, array $froms = null, array $tos = null): ?Transition
+    {
+        $marking = $this->getMarking($subject);
+
+        $result = array_filter(
```
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php:242]: I guess you were right, this is not really possible to have many transitions with the same name enabled at the very same time.  
I'm sorry for loosing your time, but could you revert this change to the one you had it yesterday? the previous version was more readable :)

Thanks

File: src/Symfony/Component/Workflow/Tests/WorkflowBuilderTrait.php (hunk 8 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Tests/WorkflowBuilderTrait.php]: done !

File: src/Symfony/Component/Workflow/Workflow.php (hunk 9 above)
Comment (epitre) [reply src/Symfony/Component/Workflow/Workflow.php]: done !

**Review by lyrixx — APPROVED** [review]
Thanks for your first contributions :+1: 

I forgot to ask: Could you update both CHANGELOG and this will be perfect :yellow_heart:

**Comment by epitre** [comment]
@lyrixx i updated the workflow changelog.
What is the other changelog you want me to update ?

**Comment by lyrixx** [comment]
The one in the bridge https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/CHANGELOG.md

**Comment by epitre** [comment]
@lyrixx Done !

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @epitre.
